File: dec2/two.mjs
import fs from "fs/promises";

function order (nums) {
    let count = 0;
    for(let i = 0; i < nums.length - 1; i++){
        if(nums[i] > nums[i + 1]) count--;
        else if(nums[i] < nums[i + 1]) count++;
    }
    return count > 0;
}

function isSafe(nums) {
    let copy = nums.slice();
    if(order(nums)) copy = nums.slice().reverse();
    let safe = true;
    for (let i = 0; i < copy.length - 1; i++) {
        let a = copy[i], b = copy[i + 1];
        let diff = a - b;
        if (diff < 1 || diff > 3) safe = false;
    }
    return safe;
}

(async function () {
    let file = await fs.readFile("./input.txt", "utf-8");
    let count = 0
    file.trim().split("\n").forEach((line, i) => {
        let nums = line.split(" ").map(Number);
        let safe = isSafe(nums);
        if (!safe){
            let counted = false;
            for(let i = 0; i < nums.length; i++){
                let copy = nums.slice();
                copy.splice(i, 1);
                let safe = isSafe(copy);
                if (safe){
                    counted = true;
                    break;
                }
            }
            if(counted){
                count++;
            }
        }
        else count++;
    });
    console.log(count);
})()

File: dec2/two.rs
use std::fs;

fn order(nums: &[i32]) -> bool {
    nums.windows(2).map(|w| w[1].cmp(&w[0]) as i32).sum::<i32>() > 0
}

fn is_safe(nums: &Vec<i32>) -> bool {
    let mut copy = nums.clone();
    if order(nums) {
        copy.reverse();
    }
    copy.windows(2).all(|w| (1..=3).contains(&(w[0] - w[1])))
}

fn main() {
    let file = fs::read_to_string("./input.txt").unwrap();
    let mut count = 0;

    for line in file.trim().lines() {
        let nums: Vec<i32> = line.split_whitespace().map(|x| x.parse().unwrap()).collect();
        if !is_safe(&nums) {
            for i in 0..nums.len() {
                let mut copy = nums.clone();
                copy.remove(i);
                if is_safe(&copy) {
                    count += 1;
                    break;
                }
            }
        } else {
            count += 1;
        }
    }

    println!("{}", count);
}
